db/snapshot/backup_task.cc

```
@@ -48,49 +48,57 @@ tasks::is_abortable backup_task_impl::is_abortable() const noexcept {
-        // with the _uploads gate.
+        // with the uploads gate.
         // Parallelism is implicitly controlled in two ways:
         //  - s3::client::claim_memory semaphore
         //  - http::client::max_connections limitation
         // FIXME -- s3::client is not abortable yet, but when it will be, need to
         // propagate impl::_as abort requests into upload_file's fibers
-        std::ignore = _client->upload_file(component_name, destination).handle_exception([comp = component_name] (auto ex) {
-            snap_log.error("Error uploading {}: {}", comp.native(), ex);
-            std::rethrow_exception(ex);
+        std::ignore = _client->upload_file(component_name, destination).handle_exception([comp = component_name, &ex] (std::exception_ptr e) {
+            snap_log.error("Error uploading {}: {}", comp.native(), e);
+            // keep the first exception
+            if (!ex) {
+                ex = std::move(e);
+            }
         }).finally([gh = std::move(gh)] {});
-        thread::maybe_yield();
-        utils::get_local_injector().inject("backup_task_pause", [] (auto& handler) {
+        co_await coroutine::maybe_yield();
+        co_await utils::get_local_injector().inject("backup_task_pause", [] (auto& handler) {
             snap_log.info("backup task: waiting");
             return handler.wait_for_message(db::timeout_clock::now() + std::chrono::minutes(2));
-        }).get();
-        impl::_as.check();
+        });
+        if (impl::_as.abort_requested()) {
+            ex = impl::_as.abort_requested_exception_ptr();
+            break;
+        }
+    }
+
+    co_await snapshot_dir_lister.close();
+    co_await uploads.close();
+    if (ex) {
+        co_await coroutine::return_exception_ptr(std::move(ex));
     }
 }
 
 future<> backup_task_impl::run() {
     co_await _snap_ctl.run_snapshot_list_operation([this] {
-        return async([this] {
-            do_backup();
```

Comment says

>  but the exceptional future is ignored by the caller of `backup_task_impl::do_backup()`, because this member function was not a coroutine. 

Does it mean that any code like

```c++
    co_await async([] {
        foo();
    });
```

ignores any exceptions thrown by `foo()`?

---

no. lemme revise the commit message so it's more accurate.